source/error.c

```diff
@@ -0,0 +1,135 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/error.h>
+#include <aws/common/common.h>
+#include <assert.h>
+
+AWS_THREAD_LOCAL int last_error = 0;
+
+static aws_error_handler global_handler = NULL;
+static void *global_error_context = NULL;
+
+AWS_THREAD_LOCAL aws_error_handler thread_handler = NULL;
+AWS_THREAD_LOCAL void *thread_handler_context = NULL;
+
+#ifndef AWS_MAX_ERROR_SLOTS
+#define AWS_MAX_ERROR_SLOTS 10
+#endif
+
+#define AWS_ERROR_SLOT_SIZE 1000
+
+static volatile const struct aws_error_info_list *error_slots[AWS_MAX_ERROR_SLOTS] = { 0 };
+
+int aws_last_error(void) {
+    return last_error;
+}
+
+static const struct aws_error_info *get_error_by_code(int err) {
+    if(err >= AWS_MAX_ERROR_SLOTS * AWS_ERROR_SLOT_SIZE) {
+        return NULL;
+    }
+
+    int slot_index = err / AWS_ERROR_SLOT_SIZE;
+    int error_index = err % AWS_ERROR_SLOT_SIZE;
+
+    volatile const struct aws_error_info_list *error_slot = error_slots[slot_index];
```

Comment:
This doesn't need to be volatile (the array at L34 does however).

Reply:
ok

Reply:
according to compiler rules it does.

Reply:
initialization discards ‘volatile’ qualifier from pointer target type [-Werror=discarded-qualifiers]

Reply:
Hmm, in that case we put volatile at the wrong point in the declaration. That should be a 

```
static const struct aws_error_info_list *volatile error_slots[AWS_MAX_ERROR_SLOTS] = { 0 };
```

That is, we want the _pointer to_ the structure to be volatile, but the structure itself, being a compile time constant, need not be volatile.

Reply:
ah.... pesky qualifiers. That worked.